runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -0,0 +1,164 @@
+local M = {}
+
+local semantic_tokens = setmetatable({}, {
+  __index = function(table, key)
+    local rtn = {}
+    rawset(table, key, rtn)
+    return rtn
+  end,
+})
+local last_tick = {}
+local active_requests = {}
+
+---@private
+local function get_bit(n, k)
+  --todo(theHamsta): remove once `bit` module is available for non-LuaJIT
+  if _G.bit then
+    return _G.bit.band(_G.bit.rshift(n, k), 1)
+  else
+    return (n / math.pow(2, k)) % 2
+  end
+end
+
+---@private
+local function modifiers_from_number(x, modifiers_table)
+  local modifiers = {}
+  for i = 0, #modifiers_table - 1 do
+    local bit = get_bit(x, i)
+    if bit == 1 then
+      table.insert(modifiers, 1, modifiers_table[i + 1])
+    end
+  end
+
+  return modifiers
+end
+
+--- |lsp-handler| for the method `textDocument/semanticTokens/full`
+---
+--- This function can be configured with |vim.lsp.with()| with the following options for `config`
+---
+--- `on_token`: A function with signature `function(ctx, token)` that is called
+---             whenever a semantic token is received from the server from context `ctx`
+---             (see |lsp-handler| for the definition of `ctx`). This can be used for highlighting the tokens.
+---             `token` is a table:
+---
+--- <pre>
+---   {
+---         line             -- line number 0-based
+---         start_char       -- start character 0-based
+---         length           -- length in characters of this token
+---         type             -- token type as string (see https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide#semantic-token-classification)
+---         modifiers        -- token modifier as string (see https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide#semantic-token-classification)
+---         offset_encoding  -- offset encoding used by the language server (see |lsp-sync|)
+---   }
+--- </pre>
+---
+--- `on_invalidate_range`: A function with signature `function(ctx, line_start, line_end)` called whenever tokens
+---                        in a specific line range (`line_start`, `line_end`) should be considered invalidated
+---                        (see |lsp-handler| for the definition of `ctx`). `line_end` can be -1 to
+---                        indicate invalidation until the end of the buffer.
+function M.on_full(err, response, ctx, config)
+  if not ctx.bufnr or not ctx.client_id then
+    return
+  end
+  active_requests[ctx.bufnr] = false
+  if config and config.on_invalidate_range then
+    config.on_invalidate_range(ctx, 0, -1)
+  end
+  -- if tick has changed our response is outdated!
+  if err or not response or not response.data or last_tick[ctx.bufnr] ~= vim.api.nvim_buf_get_changedtick(ctx.bufnr) then
```

Comment:
I still am not sure I like the current invalidation logic but I'm not going to hold the PR up over it, mostly leaving as a note.

Reply:
I don't like it neither, but I didn't now how to do it so I took the `last_tick` logic from the codelens implementation. A fix would probably require more changes in other parts of the code so I'd prefer to do it a follow up.

I tried to stay conservative and always invalidated everything whenever receiving an invalid request.

Reply:
Yeah I agree, my bigger issue is with generally how we are doing invalidation, which is why I don't want to hold this up, maybe let's file an issue about it. My general idea is we should do request invalidation (cancelling) coupled with response invalidation, and we should probably expose in context the buffer tick when the request was sent.

Reply:
Maybe add a FIXME comment to not forget it?

Reply:
Or we fix it before merging this PR... Will add a FIXME anyways

Reply:
Like I said, I'm fine merging this PR as I would rather keep the changes scoped to adding the basic version of semantic token support anyways, we can do codelens + semantic tokens invalidation in a follow-up